### ISSUE TITLE
Reindex after or during log compaction

### ISSUE THREAD

**Issue by staltz** [ssb-ngi-pointer/jitdb#199]
async-append-only-log now supports `log.compact()` https://github.com/ssb-ngi-pointer/async-append-only-log/issues/48 and we should reindex the relevant portions of jitdb indexes.

**Comment by staltz** [comment]
@arj03 I'm working on this already, but I bumped into a design issue. My original plan is to rebuild all indexes while log compaction is still ongoing, based on what [Dominic suggested here](https://github.com/ssb-ngi-pointer/ssb-db2/issues/306#issuecomment-1034479270)

> Instead of writing a separate second log, superimpose the compacted log on top of the old one, keeping track of the min uncompacted sequence number. When you do a query from the old indexes, drop results with offsets smaller than the uncompacted sequence, and mix with results from the new indexes. merge index responses by sequence numbers, so numbers smaller than the uncompacted sequence are from the compacted section and the numbers greater are from the uncompacted log.

For instance, my plan is to create a folder `.ssb/db2/indexes/_compacting` where the new index files exist, and once those are done catching up with the log, we delete all jitdb indexes in `.ssb/db2/indexes` and move up the files in `.ssb/db2/indexes/_compacting` to `.ssb/db2/indexes`.

However, does this go against the JIT design of JITDB?

Alternatively, should we just delete all JITDB indexes once log compaction is done, and let the JIT nature take care of rebuilding them?

**Comment by arj03** [comment]
Will get back to you on this tomorrow :)

**Comment by staltz** [comment]
@arj03 Thanks!

`reindex()` currently doesn't rebuild the core indexes (seq, sequence, timestamp), is there a reason for that? I think with compaction we would also have to rebuild the core indexes.

Why do you say that level indexes should be allowed to lookup while jitdb shouldn't? Is it because of speed of rebuilding the indexes? (By the way, in my tests, I felt that some jitdb indexes do take a considerable time to build, like `value_author.32prefix`, which is 6 MB in a production database, and has to write 4 bytes for each record.)

**Comment by arj03** [comment]
> reindex() currently doesn't rebuild the core indexes (seq, sequence, timestamp), is there a reason for that? I think with compaction we would also have to rebuild the core indexes.

Right, that is because for encrypted they don't change. Here they do :)

> Why do you say that level indexes should be allowed to lookup while jitdb shouldn't?

Just because it is easier. We don't need to flag to jitdb that says: please don't update indexes, just run on what you have.

I can maybe check the value_author prefix building later later to see if there is anything we can do to optimize that one.

**Comment by staltz** [comment]
Okay thanks, so I'm preparing for the plan to "not allow queries while compaction is running", and here are some disorganized thoughts (mostly questions to myself):

What happens when a jitdb query is ongoing and compaction starts? Do we cancel the query, or do we pause it and re-run it after compaction (and reindexing) is done? What if it was a 4th call to paginate and the starting `seq` would now mean that the 4th paginate would get results that were in the 3rd paginate (or, worse, would skip some results). 

This is a bit similar to questions about log.get and log.stream: do we abort it or do we wait for compaction to end and rerun it?

**Comment by arj03** [comment]
Yep agree. Also for most of these indexes you wouldn't need to do a full rebuild. Only from compaction start and onwards.

**Comment by arj03** [comment]
> I felt that some jitdb indexes do take a considerable time to build, like `value_author.32prefix`, which is 6 MB in a production database, and has to write 4 bytes for each record.)

I had to check this. Building author or type is around the same amount of time (40s). Note this is from totally empty indexes folder, so includes keys and base. Rebuilding author after type is 6s. Rebuilding both are 45s. So a tiny bit less. I tried disabling writing the jitdb indexes and it only took rebuilding from 45s to 43.2s. Decrypting is still very heavy as documented [here](https://github.com/ssb-ngi-pointer/ssb-db2/pull/318) at around 10s. If I leave the canDecrypt file and level indexes the time to build both goes down to 18.7s.

**Comment by staltz** [comment]
Super weird bug with GitHub where I can't edit the original post TODO list, so I'm copying it down here:

- [x] AAOL: compactionProgress() emits {done:true} on AAOL init
- [x] AAOL: compactionProgress() emits {done:false} immediately when compact starts
- [x] JITDB: if we start a query while log.compactionProgress() is undone, queue the query inputs
- [x] JITDB: once log.compactionProgress() emits "done" release the queue
- [ ] JITDB: reindex() should rebuild core indexes too
- [ ] JITDB: indexingActive() obz API
- [ ] JITDB: queriesActive() obz API
- [ ] DB2: reindexOffset should pass pValue to processRecord
- [ ] DB2: internal boolean levelIndexingActive
- [ ] DB2: compact() API which checks jitdb.indexingActive() and jitdb.queriesActive() and levelindexingActive and postpones itself until they are inactive
- [ ] DB2: once indexing is inactive, run log.compact() and once that's done, reindex jitdb and then run a new log.stream to reindex leveldb

**Comment by staltz** [comment]
@arj03 I think I hit a pretty sad issue: it seems like we have to reset all leveldb indexes after compact happens, and that's because most leveldb indexes hold state, and we don't know how to undo the state only for the deleted records.

Consider e.g. the about-self index which has key/value `feedId => hydratedAboutObj`. Suppose we deleted an about message with the `name` field, but we still have an about message with the `image` field. There's nothing in this index that tells us how to "partially" delete the `hydratedAboutObj`, so either we get the wrong reduced state, or we have to do a full rebuild.

Any ideas about this?

**Comment by arj03** [comment]
@staltz right, reduce based indexes versus pure. I think we could introduce that abstraction and then you'd only have to do a full reindex on the reduce based ones. Most base indexes are not reduce.

**Comment by staltz** [comment]
Yeah, we could do that split. But it might still be hard to find and remove the outdated entries. Like the EBT index, I think it's `[feedId, sequence] => offset`, we need to look up the old ones via the right-hand side, i.e. the value, not the key.

**Comment by arj03** [comment]
Oh right, you are correct. I guess there isn't any other good way than to reindex everything in this case :(
